apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Logging/ILoggerExtensions.cs

```diff
@@ -0,0 +1,98 @@
+        /// by the new ones on the returned <see cref="ILogger"/>.
+        /// </remarks>
+        public static ILogger WithLabels(this ILogger logger, params KeyValuePair<string, string>[] labels) =>
+            GaxPreconditions.CheckNotNull(logger, nameof(logger)) switch
+            {
+                AugmentedLogger augmented => augmented.WithLabels(labels),
+                _ => AugmentedLogger.ForLabels(logger, labels)
+            };
+
+        /// <summary>
+        /// Adds labels to the returned logger that may be included with every subsequent log.
+        /// How this information is included will depend on the actual <paramref name="logger"/>.
+        /// GoogleLogger(TODO: change this to a cref when I've moved it to this package) will include this
+        /// information in <see cref="LogEntry.Labels"/>.
+        /// </summary>
+        /// <remarks>
+        /// If <paramref name="logger"/> had already been augmented with labels, <paramref name="labels"/>
+        /// will be added to the existing ones on the returned <see cref="ILogger"/>
+        /// </remarks>
+        public static ILogger WithAddedLabels(this ILogger logger, params KeyValuePair<string, string>[] labels) =>
+            GaxPreconditions.CheckNotNull(logger, nameof(logger)) switch
+            {
+                AugmentedLogger augmented => augmented.WithAddedLabels(labels),
+                _ => AugmentedLogger.ForLabels(logger, labels)
+            };
+
+        private class AugmentedLogger : ILogger
+        {
+            private ILogger _innerLogger;
+            private IEnumerable<KeyValuePair<string, string>> _labels;
+
+            private AugmentedLogger(ILogger logger, IEnumerable<KeyValuePair<string, string>> labels) =>
+                (_innerLogger, _labels) = 
+                (GaxPreconditions.CheckNotNull(logger, nameof(logger)), GaxPreconditions.CheckNotNull(labels, nameof(labels)));
+
+            public static AugmentedLogger ForLabels(ILogger logger, IEnumerable<KeyValuePair<string, string>> labels) =>
+                new AugmentedLogger(logger, labels);
+
+            public AugmentedLogger WithLabels(IEnumerable<KeyValuePair<string, string>> labels) =>
+                new AugmentedLogger(_innerLogger, labels);
+
+            public AugmentedLogger WithAddedLabels(IEnumerable<KeyValuePair<string, string>> extraLabels) =>
+                new AugmentedLogger(_innerLogger, _labels.Concat(GaxPreconditions.CheckNotNull(extraLabels, nameof(extraLabels))));
+
+            public IDisposable BeginScope<TState>(TState state) => _innerLogger.BeginScope(state);
+
+            public bool IsEnabled(LogLevel logLevel) => _innerLogger.IsEnabled(logLevel);
+
+            public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+            {
+                using(_innerLogger.BeginScope(new LabellingScopeState(_labels)))
```

Comment:
Nit: space after (

Reply:
This will make the labels the *innermost* scope for logging... whereas if I've added them at the start of the program, I think I'd expect them to be the *outermost* scope. Is that basically infeasible? Or have I misunderstood?

Reply:
For the `GoogleLogger`, that this is implemented with scopes, is just that, an implementation detail (basically the only way I could think of doing this). Labels are not treated as any other scopes, their values are just added to `LogEntry.Labels` and for that the only relevant thing is the order in which the labels are added, and that is respected whether we begin the scope at logging or elsewhere.

For other loggers, I would say the same, users are not adding a scope , so they shouldn't expect labels to behave as a scope. We (I) have decided to communicate the labelling information to other loggers "in the form of a scope **on logging**" (basically because there's nothing else I can do) and loggers will treat that as they treat other scopes. I think I can be more explicit about what happens with other loggers in docs, but I think the decision can be regarded as a design decision.

Beginning the scope when the labels are added is not infeasible but would add many other consecuenses as:
- The inner logger will have an open scope for everyone else using it from the moment you add labels. That would be unexpected I think
- I'd need to return an `IDisposable` somehow (or something wrapping that) so that user code can remove the labels from the inner logger.

The main feature this is trying to solve is being able to add labels per log entry. If you want to add labels that are applied to all log entries you can use `LoggerOptions.Labels` or `ILogEntryLabelProvider` which will only afect Google's LogEntry.

Reply:
Okay - sounds like we can regard it as unfortunate but basically infeasible to get what we might want.